datman/xnat.py

```diff
@@ -88,260 +89,399 @@ class xnat(object):
-                                "permission on the xnat project?"
-                                .format(project))
-
-        return(result['items'][0])
-
-    def get_sessions(self, study):
-        logger.debug('Querying xnat server for sessions in study: {}'
-                     .format(study))
-        if not self.get_project(study):
-            raise XnatException('Invalid xnat project: {}'
-                                .format(study))
-
-        url = '{}/data/archive/projects/{}/subjects/'.format(self.server,
-                                                             study)
+            logger.debug("No projects found on server {}".format(self.server))
+            return []
+
+        if not project:
+            return result["ResultSet"]["Result"]
+
+        return result["items"]
+
+    def find_project(self, subject_id, projects=None):
+        """Find the project a subject belongs to.
+
+        Args:
+            subject_id (:obj:`str`): The subject to search for.
+            projects (:obj:`list`, optional): A list of projects to restrict
+                the search to. Defaults to None.
+
+        Returns:
+            str: The name of the XNAT project the subject belongs to. Note:
+                if the same ID is found in more than one project only the
+                first match is returned.
+        """
+        if not projects:
+            projects = self.get_projects()
+            projects = [p["ID"] for p in projects]
+
+        for project in projects:
+            if subject_id in self.get_subject_ids(project):
+                logger.debug("Found session {} in project {}"
+                             .format(subject_id, project))
+                return project
+
+    def get_subject_ids(self, project):
+        """Retrieve the IDs for all subjects within an XNAT project.
+
+        Args:
+            project (:obj:`str`): The 'Project ID' for a project on XNAT.
+
+        Raises:
+            XnatException: If the project does not exist or access fails.
+
+        Returns:
+            list: A list of string subject IDs found within the project.
+        """
+        logger.debug("Querying xnat server {} for subjects in project {}"
+                     .format(self.server, project))
+
+        if not self.get_projects(project):
+            raise XnatException("Invalid XNAT project: {}".format(project))
+
+        url = "{}/data/archive/projects/{}/subjects/".format(
+            self.server, project)
+
         try:
             result = self._make_xnat_query(url)
         except Exception:
-            raise XnatException("Failed getting xnat sessions with url: {}"
+            raise XnatException("Failed getting xnat subjects with URL {}"
                                 .format(url))
 
         if not result:
-            raise XnatException('No sessions found for study: {}'.format(study))
+            return []
+
+        return [item.get("label") for item in result["ResultSet"]["Result"]]
```

Comment:
```suggestion
        return [item["label"] for item in result["ResultSet"]["Result"]]
```
To keep consistent with lookup usage w/result, afaik .get is used moreso for the second parameter (unless item is not a dict, in which case ignore me :speak_no_evil: )

Reply:
Ah the reason I did it that way is that XNAT (appears) to always return a dictionary with the keys ResultSet and Result but the content of "Result" doesn't necessarily always have the label defined. I'm not 100% certain about that though (XNAT doesnt really document return json structure and isnt always consistent). I'll change it and catch/reraise KeyErrors as XNATExceptions for consistency